### ISSUE TITLE
pgloader fails on Greenplum querying pg_enum table

### ISSUE THREAD

**Issue by DrDonk** [dimitri/pgloader#1030]
Using the latest version of pgloader fails for Greenplum 5 and 6. It would also fail with Postgres versions before 9.1. The query of custom SQL types references a column in pg_enum that does not exist in versions of Postgres unil 9.1 and not in Redshift and Greenplum which are based primarily on Postgres 8. The column is **enumsortorder** and you can see the implementation changes in the following links to the online documentation.

[https://gpdb.docs.pivotal.io/6-0/ref_guide/system_catalogs/pg_enum.html](url)
[https://www.postgresql.org/docs/current/catalog-pg-enum.html](url)
[https://www.postgresql.org/docs/9.1/catalog-pg-enum.html](url)

Is it possible to modify the SQL query to remove the use of enumsortorder?

The relevant code links are:
list-all-sqltypes.sql [https://github.com/dimitri/pgloader/blob/fc3a1949f74bbfbebdbc023d6e55dc15e5d6df33/src/pgsql/sql/list-all-sqltypes.sql#L8](url)

pgsql-schema.lisp [https://github.com/dimitri/pgloaderpgsql-schema.lisp /blob/b8da7dd2e9c2de281af6b8e6dfaef8100212dfab/src/pgsql/pgsql-schema.lisp#L425](url)

pgsql-schema.lisp [[https://github.com/dimitri/pgloader/blob/b8da7dd2e9c2de281af6b8e6dfaef8100212dfab/src/pgsql/pgsql-schema.lisp#L27](url)

  - [ ] pgloader --version
pgloader version "3.6.1"
compiled with SBCL 1.5.6

  - [ ] did you test a fresh compile from the source tree?
Yes using `brew install --HEAD pgloader`
    
  - [ ] did you search for other similar issues?
  
Yes

  - [ ] how can I reproduce the bug?
```
LOAD CSV
     FROM INLINE
     (
        Col1,
        Col2
     )
     INTO postgresql://<redacted>:5432/postgres?"csv-numeric"
     (
        Col1,
        Col2
     )

     WITH truncate,
        skip header = 1,
        fields optionally enclosed by '"',
        fields escaped by double-quote,
        fields terminated by ','

    BEFORE LOAD DO
        $$ drop table if exists "csv-numeric"; $$,
        $$ CREATE TABLE "csv-numeric"
           (
                Col1 text,
                Col2 numeric
           )
        $$;

Col1,Col2
Val1,-123
Val2,456.999
Val3,789
Val4,5.8E-11
Val5,1.2345E-08
Val6,1.2345E08
```

  - [ ] pgloader output you obtain
  
```
pgloader version 3.6.1
sb-impl::*default-external-format* :UTF-8
tmpdir: #P"/var/folders/4m/f0cw5lp14vs22w_n1d3s889h0000gn/T/"
2019-09-24T13:25:58.005000+01:00 NOTICE Starting pgloader, log system is ready.
2019-09-24T13:25:58.009000+01:00 INFO Starting monitor
2019-09-24T13:25:58.011000+01:00 LOG pgloader version "3.6.1"
2019-09-24T13:25:58.011000+01:00 LOG Data errors in '/private/tmp/pgloader/'
2019-09-24T13:25:58.011000+01:00 LOG Parsing commands from file #P"/Users/dave/Plaid/Tartan/jira/plaid-564/input/pgload-gp-issue.load"
2019-09-24T13:25:58.048000+01:00 INFO Parsed command:
LOAD CSV
     FROM INLINE
     (
        Col1,
        Col2
     )
     INTO postgresql://<redacted>:5432/postgres?"csv-numeric"
     (
        Col1,
        Col2
     )

     WITH truncate,
        skip header = 1,
        fields optionally enclosed by '"',
        fields escaped by double-quote,
        fields terminated by ','

    BEFORE LOAD DO
        $$ drop table if exists "csv-numeric"; $$,
        $$ CREATE TABLE "csv-numeric"
           (
                Col1 text,
                Col2 numeric
           )
        $$;


2019-09-24T13:25:58.289000+01:00 DEBUG CONNECTED TO #<PGLOADER.PGSQL:PGSQL-CONNECTION pgsql://<redacted>:5432/postgres {1005D273D3}>
2019-09-24T13:25:58.289000+01:00 DEBUG SET client_encoding TO 'utf8'
2019-09-24T13:25:58.531000+01:00 DEBUG SET application_name TO 'pgloader'
2019-09-24T13:25:58.981000+01:00 NOTICE Executing SQL block for before load
2019-09-24T13:25:59.222000+01:00 DEBUG CONNECTED TO #<PGLOADER.PGSQL:PGSQL-CONNECTION pgsql://<redacted>:5432/postgres {1005D273D3}>
2019-09-24T13:25:59.222000+01:00 DEBUG SET client_encoding TO 'utf8'
2019-09-24T13:25:59.365000+01:00 DEBUG SET application_name TO 'pgloader'
2019-09-24T13:25:59.911000+01:00 DEBUG BEGIN
2019-09-24T13:26:00.048000+01:00 SQL drop table if exists "csv-numeric";
2019-09-24T13:26:00.506000+01:00 SQL CREATE TABLE "csv-numeric"
           (
                Col1 text,
                Col2 numeric
           )
2019-09-24T13:26:01.407000+01:00 DEBUG CONNECTED TO #<PGLOADER.PGSQL:PGSQL-CONNECTION pgsql://<redacted>:5432/postgres {1005D273D3}>
2019-09-24T13:26:01.407000+01:00 DEBUG SET client_encoding TO 'utf8'
2019-09-24T13:26:01.549000+01:00 DEBUG SET application_name TO 'pgloader'
2019-09-24T13:26:02.099000+01:00 SQL --
-- get user defined SQL types
--
  select nt.nspname,
         extname,
         typname,
         case when enum.enumtypid is not null
              then array_agg(enum.enumlabel order by enumsortorder)
          end as enumvalues

    from pg_class c
         join pg_namespace n on n.oid = c.relnamespace
         left join pg_attribute a on c.oid = a.attrelid and a.attnum > 0
         join pg_type t on t.oid = a.atttypid
         left join pg_namespace nt on nt.oid = t.typnamespace
         left join pg_depend d on d.classid = 'pg_type'::regclass
                              and d.refclassid = 'pg_extension'::regclass
                              and d.objid = t.oid
         left join pg_extension e on refobjid = e.oid
         left join pg_enum enum on enum.enumtypid = t.oid

   where nt.nspname !~ '^pg_' and nt.nspname <> 'information_schema'
         and n.nspname !~ '^pg_' and n.nspname <> 'information_schema'
         and c.relkind in ('r', 'f', 'p')
           and ((n.nspname = 'public' and c.relname = 'csv-numeric'))
           
         and
           (   t.typrelid = 0
            or
               (select c.relkind = 'c'
                 from pg_class c
                where c.oid = t.typrelid)
           )
           and not exists
             (
                select 1
                  from pg_type el
                 where el.oid = t.typelem
                   and el.typarray = t.oid
              )

group by nt.nspname, extname, typname, enumtypid
order by nt.nspname, extname, typname, enumtypid;
2019-09-24T13:26:02.174000+01:00 ERROR Database error 42703: column "enumsortorder" does not exist
QUERY: --
-- get user defined SQL types
--
  select nt.nspname,
         extname,
         typname,
         case when enum.enumtypid is not null
              then array_agg(enum.enumlabel order by enumsortorder)
          end as enumvalues

    from pg_class c
         join pg_namespace n on n.oid = c.relnamespace
         left join pg_attribute a on c.oid = a.attrelid and a.attnum > 0
         join pg_type t on t.oid = a.atttypid
         left join pg_namespace nt on nt.oid = t.typnamespace
         left join pg_depend d on d.classid = 'pg_type'::regclass
                              and d.refclassid = 'pg_extension'::regclass
                              and d.objid = t.oid
         left join pg_extension e on refobjid = e.oid
         left join pg_enum enum on enum.enumtypid = t.oid

   where nt.nspname !~ '^pg_' and nt.nspname <> 'information_schema'
         and n.nspname !~ '^pg_' and n.nspname <> 'information_schema'
         and c.relkind in ('r', 'f', 'p')
           and ((n.nspname = 'public' and c.relname = 'csv-numeric'))
           
         and
           (   t.typrelid = 0
            or
               (select c.relkind = 'c'
                 from pg_class c
                where c.oid = t.typrelid)
           )
           and not exists
             (
                select 1
                  from pg_type el
                 where el.oid = t.typelem
                   and el.typarray = t.oid
              )

group by nt.nspname, extname, typname, enumtypid
order by nt.nspname, extname, typname, enumtypid;
2019-09-24T13:26:02.244000+01:00 FATAL Failed to prepare target PostgreSQL table.
2019-09-24T13:26:02.245000+01:00 FATAL Database error 42703: column "enumsortorder" does not exist
QUERY: --
-- get user defined SQL types
--
  select nt.nspname,
         extname,
         typname,
         case when enum.enumtypid is not null
              then array_agg(enum.enumlabel order by enumsortorder)
          end as enumvalues

    from pg_class c
         join pg_namespace n on n.oid = c.relnamespace
         left join pg_attribute a on c.oid = a.attrelid and a.attnum > 0
         join pg_type t on t.oid = a.atttypid
         left join pg_namespace nt on nt.oid = t.typnamespace
         left join pg_depend d on d.classid = 'pg_type'::regclass
                              and d.refclassid = 'pg_extension'::regclass
                              and d.objid = t.oid
         left join pg_extension e on refobjid = e.oid
         left join pg_enum enum on enum.enumtypid = t.oid

   where nt.nspname !~ '^pg_' and nt.nspname <> 'information_schema'
         and n.nspname !~ '^pg_' and n.nspname <> 'information_schema'
         and c.relkind in ('r', 'f', 'p')
           and ((n.nspname = 'public' and c.relname = 'csv-numeric'))
           
         and
           (   t.typrelid = 0
            or
               (select c.relkind = 'c'
                 from pg_class c
                where c.oid = t.typrelid)
           )
           and not exists
             (
                select 1
                  from pg_type el
                 where el.oid = t.typelem
                   and el.typarray = t.oid
              )

group by nt.nspname, extname, typname, enumtypid
order by nt.nspname, extname, typname, enumtypid;
2019-09-24T13:26:02.245000+01:00 LOG report summary reset
       table name     errors       read   imported      bytes      total time       read      write
-----------------  ---------  ---------  ---------  ---------  --------------  ---------  ---------
            fetch          0          0          0                     0.006s    
      before load          0          2          2                     2.171s    
-----------------  ---------  ---------  ---------  ---------  --------------  ---------  ---------
-----------------  ---------  ---------  ---------  ---------  --------------  ---------  ---------
2019-09-24T13:26:02.275000+01:00 INFO Stopping monitor
```
  - [ ] How the data is different from what you expected, if relevant
No data loaded due to query errors.

**Comment by DrDonk** [comment]
The recently released Greenplum 6.0 does work as the column is present now. It does not work with previous versions.

**Comment by dimitri** [comment]
Those version of Postgres are not supported anymore, and I'm happy to see that newer version of Greenplum have the `enumsortorder`. Now you could send a patch to add support for Redshift by adding a derivative of the SQL query used that works there, and place it in `src/pgsql/sql/8.0` as an SQL variant for version 8.0.
